ateam_ai/src/util/directed_graph.hpp

```diff
@@ -0,0 +1,82 @@
+// Copyright 2021 A Team
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL
+// THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+#ifndef UTIL__DIRECTED_GRAPH_HPP_
+#define UTIL__DIRECTED_GRAPH_HPP_
+
+#include <map>
+#include <vector>
+
+/**
+ * Directed graph of nodes with no restriction on connectivity
```

Comment:
Nit: as written this is a DAG, so there is a restriction on connectivity. I assume that's intentional because we really do not want loops in dependency graphs, but I'd mention it in the comment (unless we want to allow for the ability to modify the set of parents of a node after it's been created, but I'd prefer we _don't_ add that functionality for exactly this reason).